go/pkg/client/cas.go

```diff
@@ -722,53 +904,97 @@ func (c *Client) DownloadFiles(ctx context.Context, execRoot string, outputs map
 		}
 	}
 
-	eg, eCtx := errgroup.WithContext(ctx)
+	ctx := context.Background()
 	for i, batch := range batches {
-		i, batch := i, batch     // https://golang.org/doc/faq#closures_and_goroutines
-		c.casDownloaders <- true // Reserve an downloader thread.
-		eg.Go(func() error {
-			defer func() { <-c.casDownloaders }()
+		i, batch := i, batch // https://golang.org/doc/faq#closures_and_goroutines
+		go func() (err error) {
+			c.casDownloaders.Acquire(ctx, 1)
+			defer c.casDownloaders.Release(1)
 			if i%logInterval == 0 {
 				log.V(2).Infof("%d batches left to download", len(batches)-i)
 			}
 			if len(batch) > 1 {
 				log.V(3).Infof("Downloading batch of %d files", len(batch))
-				bchMap, err := c.BatchDownloadBlobs(eCtx, batch)
+				bchMap, err := c.BatchDownloadBlobs(ctx, batch)
 				if err != nil {
+					afterDownload(batch, reqs, err)
 					return err
 				}
 				for dg, data := range bchMap {
-					out := outputs[dg]
-					perm := c.RegularMode
-					if out.IsExecutable {
-						perm = c.ExecutableMode
-					}
-					if err := ioutil.WriteFile(filepath.Join(execRoot, out.Path), data, perm); err != nil {
-						return err
+					for _, r := range reqs[dg] {
+						perm := c.RegularMode
+						if r.output.IsExecutable {
+							perm = c.ExecutableMode
+						}
+						r.wait <- ioutil.WriteFile(filepath.Join(r.execRoot, r.output.Path), data, perm)
 					}
 				}
 			} else {
-				out := outputs[batch[0]]
-				path := filepath.Join(execRoot, out.Path)
-				log.V(3).Infof("Downloading single file with digest %s to %s", out.Digest, path)
-				if _, err := c.ReadBlobToFile(ctx, out.Digest, path); err != nil {
+				// We cannot release the lock after each file copy in this case, because
+				// the caller might move the file, and we don't have the contents in memory.
```

Comment:
Would it be simpler if we download the blob with the same digest multiple times? So that the download logic is keyed by the output path not the digest? Its rare enough that I think its not worth the extra logic in the code.